epsagon/events/azure.py

```diff
@@ -0,0 +1,227 @@
+            self,
+            wrapped,
+            instance,
+            args,
+            kwargs,
+            start_time,
+            response,
+            exception
+    ):
+        """
+        Initialize.
+        :param wrapped: wrapt's wrapped
+        :param instance: wrapt's instance
+        :param args: wrapt's args
+        :param kwargs: wrapt's kwargs
+        :param start_time: Start timestamp (epoch)
+        :param response: response data
+        :param exception: Exception (if happened)
+        """
+        super(AzureEvent, self).__init__(start_time)
+        self.event_id = 'azure-{}'.format(str(uuid4()))
+        self.resource['operation'] = wrapped.__name__
+        self.resource['metadata'] = {}
+
+        if response is not None:
+            self.update_response(response)
+
+        if exception is not None:
+            self.set_exception(exception, traceback.format_exc())
+
+    def update_response(self, _response):
+        """
+        Adds response data to event.
+        :param _response: Response from azure
+        :return: None
+        """
+        self.RESPONSE_TO_FUNC.get(self.resource['operation'], empty_func)()
+
+    def set_exception(self, exception, traceback_data, handled=True):
+        """
+        see {Event.set_exception}
+        """
+        super(AzureEvent, self).set_exception(
+            exception,
+            traceback_data,
+            handled=handled
+        )
+
+        # Specific handling for azure errors
+        if isinstance(exception, AzureError):
+            self.event_id = exception.headers.get('x-ms-activity-id')
+            self.resource['metadata']['azure_error'] = True
+            self.resource['metadata']['status_code'] = exception.status_code
+            self.resource['metadata']['error_message'] = exception.message
+            self.resource['metadata']['reason'] = exception.reason
+
+
+class AzureCosmosContainerEvent(AzureEvent):
+    """
+    Represents CosmosDB Container Azure event.
+    """
+
+    RESOURCE_TYPE = 'cosmos_db'
+
+    def __init__(
+            self,
+            wrapped,
+            instance,
+            args,
+            kwargs,
+            start_time,
+            response,
+            exception
+    ):
+        """
+        Initialize.
+        """
+        super(AzureCosmosContainerEvent, self).__init__(
+            wrapped,
+            instance,
+            args,
+            kwargs,
+            start_time,
+            response,
+            exception
+        )
+        self.RESPONSE_TO_FUNC.update({
+            'upsert_item': self.upsert_item_response,
+        })
+
+        self.OPERATION_TO_FUNC.update({
+            'upsert_item': self.upsert_item_request,
+            'delete_item': self.delete_item_request,
+            'query_items': self.query_items_request,
+        })
+        self.kwargs = kwargs
+        self.args = args
+        self.response = response
+
+        self.resource['metadata']['azure.location'] = (
```

Comment:
How solid is this format? consider try except, in case there are variants we didn't account for yet (e.g. no "_" so [1] will fail)

Reply:
The instrumentation in Azure case is very specific (this looks only on container operations) - so this field remains in the same structure